pkg/ddl/notifier/testkit_test.go

```diff
@@ -206,3 +208,88 @@ func TestDeliverOrderAndCleanup(t *testing.T) {
+	tk.MustExec("truncate table t")                                                                                                      // ActionTruncateTable
+	tk.MustExec("drop table t1")                                                                                                         // ActionDropTable
+	tk.MustExec("alter table t modify column a varchar(15)")                                                                             // ActionModifyColumn
+	tk.MustExec("alter table t add column b int")                                                                                        // ActionAddColumn
+	tk.MustExec("alter table t add index(b)")
+	tk.MustExec("create table t1(b int key, FOREIGN KEY (b) REFERENCES t(b) ON DELETE CASCADE);") // ActionCreateTable with foreign key
+	tk.MustExec("alter table t1 add column c int, add column d varchar(10)")                      // ActionAddColumn
+
+	ctx := context.Background()
+	s := notifier.OpenTableStore("test", "ddl_notifier")
+	se := sess.NewSession(tk.Session())
+	got, err := s.List(ctx, se)
+	require.NoError(t, err)
+	require.Len(t, got, 16)
+	rows := tk.MustQuery("select schema_change, multi_schema_change_seq from test.ddl_notifier").Rows()
+	tps := make([]model.ActionType, len(rows))
+	multiSchemaChangeSeqs := make([]int64, len(rows))
+	for i, row := range rows {
+		event := &notifier.SchemaChangeEvent{}
+		err = event.UnmarshalJSON([]byte(row[0].(string)))
+		require.NoError(t, err)
+		tps[i] = event.GetType()
+		seq, err := strconv.Atoi(row[1].(string))
+		require.NoError(t, err)
+		multiSchemaChangeSeqs[i] = int64(seq)
+	}
+	require.Equal(t, tps, []model.ActionType{
+		model.ActionCreateTable,
+		model.ActionAlterTablePartitioning,
+		model.ActionReorganizePartition,
+		model.ActionTruncateTablePartition,
+		model.ActionDropTablePartition,
+		model.ActionAddTablePartition,
+		model.ActionCreateTable,
+		model.ActionExchangeTablePartition,
+		model.ActionRemovePartitioning,
+		model.ActionTruncateTable,
+		model.ActionDropTable,
+		model.ActionModifyColumn,
+		model.ActionAddColumn,
+		model.ActionCreateTable,
+		model.ActionAddColumn,
+		model.ActionAddColumn,
+	})
+	require.Equal(t, multiSchemaChangeSeqs, []int64{-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 0, 1})
+}
+
+func TestPublishEventError(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("USE test")
+	tk.MustExec("DROP TABLE IF EXISTS ddl_notifier")
+	tk.MustExec(tableStructure)
+	notifier.DefaultStore = notifier.OpenTableStore("test", "ddl_notifier")
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/pkg/ddl/asyncNotifyEventError", "return()"))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/pkg/ddl/asyncNotifyEventError"))
+	}()
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int)")
```

Comment:
I think this test is not expected. If DDL executes successfully (`MustExec` here),  the `ddl_notifier` table must have corresponding row. Otherwise the DDL is lost to publish.

Reply:
After inject asyncNotifyEventError, the ddj job failed but without an error directly return, and the error log can be seen.
<img width="1080" alt="截屏2024-10-11 17 57 48" src="https://github.com/user-attachments/assets/0d1026e2-a73c-4430-a3dc-13e08a3bbd3f">

Reply:
I'm not sure why `MustExec` doesn't fail the test if DDL is failed 😂 Will take a look tomorrow

Reply:
> the ddj job failed

`SchemaState` is public and the job is finished. This is not expected.

```
[2024/10/14 10:18:12.000 +08:00] [INFO] [job_worker.go:378] ["finish DDL job"] [category=ddl] [jobID=113] [conn=1] [job="ID:113, Type:create table, State:synced, SchemaState:public, SchemaID:2, TableID:112, RowCount:0, ArgLen:0, start time: 2024-10-14 10:18:10.955 +0800 CST, Err:[ddl:-1]mock publish event error, ErrCount:1, SnapshotVersion:0, Version: v2"]
```

Please be careful about the meaning of test code because it reveals bug.

Reply:
Thanks, got it. I have fix the comment. And move the publish logic before finish the job, ptal again https://github.com/pingcap/tidb/pull/56425/commits/71c798a80e8d18ad85f81111bca150377d4ca362